### PR TITLE
Allow DDL changes to Compressed Hypertables via Migration

### PR DESCRIPTION
**Pull request by tylerfontaine**
To avoid having to decompress an entire hypertable to make a DDL change,
this procedure facilitates copying data to a new table with the correct
structure. It copies whole chunks at a time, and compresses them in the
new table as it goes.

### CODE REVIEW

File: utils/migrate_compressed_ddl.sql [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+--   other dependent objects. It only copies the data to a new table.
+--
+-- Usage:
+--   CALL migrate_data('copy_from', 'copy_to', '30 days');
+--
+ 
+
+
+CREATE OR REPLACE procedure migrate_data(
+	old_table text,
```
Comment (davidkohn88) [inline utils/migrate_compressed_ddl.sql:34]: I'd use regclasses for these to properly account for schema qualification

File: utils/migrate_compressed_ddl.sql [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	EXECUTE format('insert into %I
+	(select * from %s)', new_table, c_row.chunkname);
+	
+	-- get most recent chunk from the new hypertable after copying the 
+	-- whole other chunk is finished.
+	-- it will only grab chunks to drop up to the older_than interval 
+	-- specified.
+	select a.show_chunks into last_chunk from 
+	(select * from 
+	show_chunks('copyto', older_than => older_than)
```
Comment (davidkohn88) [inline utils/migrate_compressed_ddl.sql:66]: The static string in this seems wrong to me?

File: utils/migrate_compressed_ddl.sql [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+CREATE TEMPORARY TABLE chunknames (chunkname text);
+
+insert into chunknames
+(select * from show_chunks(old_table));
+
+-- loop through (one at a time) the chunks and copy the data to the new 
+-- hypertable
+FOR c_row in c_curs LOOP
+	EXECUTE format('insert into %I
+	(select * from %s)', new_table, c_row.chunkname);
```
Comment (davidkohn88) [inline utils/migrate_compressed_ddl.sql:58]: Switching to regclasses should help here, we may have to do some casting to text beforehand into other variables, but shouldn't be too bad, and then we won't be switching between %c/%I and we'll actually make sure things are formatted correctly.

File: utils/migrate_compressed_ddl.sql [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	old_table text, 
+	new_table text, 
+	older_than interval)
+language plpgsql
+as 
+$$
+DECLARE
+c_row record;
+c_curs cursor for 
+select * from chunknames;
```
Comment (davidkohn88) [inline utils/migrate_compressed_ddl.sql:43]: I think you'll have to open this later given that the temp table isn't created until later in the run, or at least it's a bit confusing.

Comment (davidkohn88) [reply utils/migrate_compressed_ddl.sql:43]: I might even just avoid the temp table all together and use an array, which you can also loop through.

File: utils/migrate_compressed_ddl.sql [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+insert into chunknames
+(select * from show_chunks(old_table));
+
+-- loop through (one at a time) the chunks and copy the data to the new 
+-- hypertable
+FOR c_row in c_curs LOOP
+	EXECUTE format('insert into %I
+	(select * from %s)', new_table, c_row.chunkname);
+	
+	-- get most recent chunk from the new hypertable after copying the
```
Comment (davidkohn88) [inline utils/migrate_compressed_ddl.sql:60]: I think this whole thing is not accounting correctly for space partitioning...also, does selecting directly from compressed chunks work or do you have to go through the hypertable in order to get proper routing?

**Review by davidkohn88 — CHANGES_REQUESTED** [review]
I think this needs a bit more work, there's some edge cases we're not accounting for. I think one thing that would be really helpful to move it forward would be to create a small test procedure after this that created the hypertable, added some data, compressed it and moved it over. Ideally a hypertable with space partitioning.

File: utils/migrate_compressed_ddl.sql (hunk 4 above)
Comment (tylerfontaine) [reply utils/migrate_compressed_ddl.sql:43]: I thought the same, but it does actually work. Though I'm not against an array. I'll just move to an array. That would mean I wouldn't need the cursor either anyway.

File: utils/migrate_compressed_ddl.sql (hunk 1 above)
Comment (tylerfontaine) [reply utils/migrate_compressed_ddl.sql:34]: Good point.

File: utils/migrate_compressed_ddl.sql (hunk 3 above)
Comment (tylerfontaine) [reply utils/migrate_compressed_ddl.sql:58]: The problem wasn't their classes so much as you have to do this as a dynamic query in order for the variables to be read in. Or at least, that was the only way that made it work.

File: utils/migrate_compressed_ddl.sql (hunk 5 above)
Comment (tylerfontaine) [reply utils/migrate_compressed_ddl.sql:60]: It does not - I somehow managed to do all this testing with an uncompressed copyfrom table. Oops

File: utils/migrate_compressed_ddl.sql (hunk 2 above)
Comment (tylerfontaine) [reply utils/migrate_compressed_ddl.sql:66]: It is. Will fix.
